compiler/rustc_errors/src/emitter.rs

```diff
@@ -297,9 +297,16 @@ pub trait Emitter: Translate {
                     // are some which do actually involve macros.
                     ExpnKind::Desugaring(..) | ExpnKind::AstPass(..) => None,
 
-                    ExpnKind::Macro(macro_kind, name) => {
-                        Some((macro_kind, name, expn_data.hide_backtrace))
-                    }
+                    ExpnKind::Macro(macro_kind, name) => Some((
+                        macro_kind,
+                        name,
+                        expn_data.hide_backtrace
+                            || expn_data.allow_internal_unstable.is_some_and(|aiu| {
+                                aiu.contains(&sym::fmt_internals)
+                                    || aiu.contains(&sym::print_internals)
+                                    || aiu.contains(&sym::liballoc_internals)
+                            }),
+                    )),
```

Comment:
Might be worth pulling everything after the first `||` out into a separate variable.

Also, I don't really understand how this works, i.e. how `allow_internal_unstable` is involved. A brief comment explaining it would be useful.